File: dsa-problems/ProbabilisticDS/solutions/python/PDS-013-xor-filters.py
import math
import sys

def xor_filter_stats(n: int, b: int):
    cells = math.ceil(1.23 * n)
    mem = int(cells * b)
    fpr = 2.0 ** -b
    return mem, fpr

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    n = int(data[0])
    b = int(data[1])
    mem, fpr = xor_filter_stats(n, b)
    print(f"{mem} {fpr:.6f}")

if __name__ == "__main__":
    main()

File: dsa-problems/ProbabilisticDS/solutions/javascript/PDS-013-xor-filters.js
const readline = require("readline");

function xorFilterStats(n, b) {
  const cells = Math.ceil(1.23 * n);
  const mem = cells * b;
  const fpr = Math.pow(2.0, -b);
  return [mem, fpr];
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  const n = parseInt(data[0], 10);
  const b = parseInt(data[1], 10);
  const res = xorFilterStats(n, b);
  console.log(res[0] + " " + res[1].toFixed(6));
});
